### DISCUSSION TITLE
How to use json for workspaces?

### DISCUSSION THREAD

**Discussion by beyond9thousand** [elkowar/eww#537] [Q&A]
I am currently using a bash script integrated through literal for handling workspaces. I am aware that using json is a better option but am unable to decipher the method to do so. Can someone help out?

**Comment by VuiMuich** [comment]
Which WM are you using?
Some, like `leftwm` have a native method (`leftwm-state`) to prind info in JSON format. Some like `qtile`, `xmonad` or `awesomewm` should be ralatively easily hackable to do.

In your widget you use `'${some-field.child-field}'` to get the value of `child-field`. You can have a look at how magic variables are handled.

**Reply by beyond9thousand** [reply]
I am using bspwm

**Reply by VuiMuich** [reply]
afaim `bspc wm -d` dumps you the current state as JSON, but I am no real bspwm user so I am not sure if this includes all the info you want. And iirc this was no continuous output, so you would need to poll for it.

**Reply by viandoxdev** [reply]
```sh
bspc wm -d; bspc subscribe desktop | while read -r; do bspw wm -d; done
```
Will give continuous output of the state of the wm in json. You can just put that in a deflisten to get a variable with it. If you want to restructure the json to be easier to use I'd recommend piping [jq](https://stedolan.github.io/jq/manual/):
```sh
_wm() { bspc wm -d | jq -<options> 'jq filter'; }; _wm; bspc subscribe desktop | while read -r; do _wm; done 
```
For the yuck you could do something like that:
```clojure
(deflisten wm :initial '...' `the command above`)
(defwidget workspace [index]
    (whatever {wm[index]})) ; Or whatever json access you need to get the data you want
(defwidget workspaces []
    (workspace :index 0)
    ; ....
    (workspace :index n))
```
Or use a for
```clojure
(defwidget workspaces []
    (for ws in {wm.workspaces} ; again adapt that to the structure of your json
        (whatever {ws.<something>})))
```

**Reply by beyond9thousand** [reply]
not quite sure about what the `jq -<options> 'jq filter'` is, it returns an error. am i supposed to substitute it with something?

Ok nvm got it. Seeing if this would work for the use case
